File: py_src/taskito/dashboard/overrides_store.py
"""Persistent task & queue runtime overrides.

Operators tune individual task or queue behaviour (rate limits, concurrency
caps, retry policy, timeouts, priority, paused state) at runtime via the
dashboard. The decorator-declared values become the *defaults* — any override
recorded here wins.

Storage layout in ``dashboard_settings``:

- ``overrides:task:<task_name>`` — JSON of overridden fields for that task
- ``overrides:queue:<queue_name>`` — JSON of overridden fields for that queue

Overrides are applied at worker startup (see
:meth:`taskito.mixins.lifecycle.QueueLifecycleMixin.start_worker`).
Changes to the store DO NOT take effect on a running worker until it is
restarted — the dashboard surfaces this so operators aren't surprised.

The contract is intentionally minimal: only the fields below can be
overridden. The store rejects anything else so a typo can't write garbage
through the dashboard.
"""

from __future__ import annotations

import json
import logging
import time
from dataclasses import asdict, dataclass
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from taskito.app import Queue


TASK_PREFIX = "overrides:task:"
QUEUE_PREFIX = "overrides:queue:"

logger = logging.getLogger("taskito.dashboard.overrides")


# ── Allowed override fields ────────────────────────────────────────────


TASK_OVERRIDE_FIELDS: frozenset[str] = frozenset(
    {
        "rate_limit",
        "max_concurrent",
        "max_retries",
        "retry_backoff",
        "timeout",
        "priority",
        "paused",
    }
)

QUEUE_OVERRIDE_FIELDS: frozenset[str] = frozenset(
    {
        "rate_limit",
        "max_concurrent",
        "paused",
    }
)


# ── Data classes ───────────────────────────────────────────────────────


@dataclass(frozen=True)
class TaskOverride:
    """An operator-set override for a registered task."""

    task_name: str
    rate_limit: str | None = None
    max_concurrent: int | None = None
    max_retries: int | None = None
    retry_backoff: float | None = None
    timeout: int | None = None
    priority: int | None = None
    paused: bool = False
    updated_at: int = 0

    def as_patch(self) -> dict[str, Any]:
        """Return a dict of only the non-default fields (those the operator
        actually set). The empty/default values are NOT patched onto the
        underlying ``PyTaskConfig`` — they continue to use the decorator
        value."""
        patch: dict[str, Any] = {}
        for field in TASK_OVERRIDE_FIELDS:
            if field == "paused":
                continue  # handled separately; not a PyTaskConfig field
            value = getattr(self, field)
            if value is not None:
                patch[field] = value
        return patch


@dataclass(frozen=True)
class QueueOverride:
    """An operator-set override for a queue."""

    queue_name: str
    rate_limit: str | None = None
    max_concurrent: int | None = None
    paused: bool = False
    updated_at: int = 0


# ── Validation ─────────────────────────────────────────────────────────


def _validate_task_fields(fields: dict[str, Any]) -> None:
    unknown = set(fields) - TASK_OVERRIDE_FIELDS
    if unknown:
        raise ValueError(f"unknown task override fields: {sorted(unknown)}")
    _validate_rate_limit(fields.get("rate_limit"))
    _validate_max_concurrent(fields.get("max_concurrent"))
    _validate_int_field(fields, "max_retries", minimum=0)
    _validate_float_field(fields, "retry_backoff", minimum=0)
    _validate_int_field(fields, "timeout", minimum=1)
    _validate_int_field(fields, "priority")
    _validate_bool_field(fields, "paused")


def _validate_queue_fields(fields: dict[str, Any]) -> None:
    unknown = set(fields) - QUEUE_OVERRIDE_FIELDS
    if unknown:
        raise ValueError(f"unknown queue override fields: {sorted(unknown)}")
    _validate_rate_limit(fields.get("rate_limit"))
    _validate_max_concurrent(fields.get("max_concurrent"))
    _validate_bool_field(fields, "paused")


def _validate_rate_limit(value: Any) -> None:
    if value is None:
        return
    if not isinstance(value, str) or not value:
        raise ValueError("rate_limit must be a non-empty string like '100/m'")
    # Cheap shape check; rate-limit parsing happens in Rust.
    if "/" not in value:
        raise ValueError("rate_limit must contain a unit, e.g. '10/s', '100/m', '3600/h'")


def _validate_max_concurrent(value: Any) -> None:
    if value is None:
        return
    if not isinstance(value, int) or isinstance(value, bool) or value < 0:
        raise ValueError("max_concurrent must be a non-negative integer")


def _validate_int_field(fields: dict[str, Any], name: str, *, minimum: int | None = None) -> None:
    value = fields.get(name)
    if value is None:
        return
    if not isinstance(value, int) or isinstance(value, bool):
        raise ValueError(f"{name} must be an integer")
    if minimum is not None and value < minimum:
        raise ValueError(f"{name} must be >= {minimum}")


def _validate_float_field(
    fields: dict[str, Any], name: str, *, minimum: float | None = None
) -> None:
    value = fields.get(name)
    if value is None:
        return
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise ValueError(f"{name} must be a number")
    if minimum is not None and value < minimum:
        raise ValueError(f"{name} must be >= {minimum}")


def _validate_bool_field(fields: dict[str, Any], name: str) -> None:
    value = fields.get(name)
    if value is not None and not isinstance(value, bool):
        raise ValueError(f"{name} must be a boolean")


# ── Store ──────────────────────────────────────────────────────────────


def _now() -> int:
    return int(time.time() * 1000)


def _parse_json(raw: str | None) -> dict[str, Any]:
    if not raw:
        return {}
    try:
        data = json.loads(raw)
    except json.JSONDecodeError:
        logger.warning("overrides entry is not valid JSON; treating as empty")
        return {}
    return data if isinstance(data, dict) else {}


class OverridesStore:
    """CRUD for per-task and per-queue runtime overrides."""

    def __init__(self, queue: Queue) -> None:
        self._queue = queue

    # ── Tasks ──────────────────────────────────────────────────

    def list_tasks(self) -> dict[str, TaskOverride]:
        """Return ``{task_name: TaskOverride}`` for every task with an override."""
        out: dict[str, TaskOverride] = {}
        for key, raw in self._queue.list_settings().items():
            if not key.startswith(TASK_PREFIX):
                continue
            task_name = key[len(TASK_PREFIX) :]
            out[task_name] = self._row_to_task(task_name, _parse_json(raw))
        return out

    def get_task(self, task_name: str) -> TaskOverride | None:
        raw = self._queue.get_setting(TASK_PREFIX + task_name)
        if not raw:
            return None
        return self._row_to_task(task_name, _parse_json(raw))

    def set_task(self, task_name: str, fields: dict[str, Any]) -> TaskOverride:
        _validate_task_fields(fields)
        if not task_name:
            raise ValueError("task_name must not be empty")
        existing = self.get_task(task_name)
        merged: dict[str, Any] = {}
        if existing is not None:
            merged.update({k: v for k, v in asdict(existing).items() if v is not None})
            merged.pop("task_name", None)
            merged.pop("updated_at", None)
        for k, v in fields.items():
            if v is None:
                merged.pop(k, None)
            else:
                merged[k] = v
        merged["updated_at"] = _now()
        self._queue.set_setting(TASK_PREFIX + task_name, json.dumps(merged, separators=(",", ":")))
        return self._row_to_task(task_name, merged)

    def clear_task(self, task_name: str) -> bool:
        return self._queue.delete_setting(TASK_PREFIX + task_name)

    @staticmethod
    def _row_to_task(task_name: str, row: dict[str, Any]) -> TaskOverride:
        return TaskOverride(
            task_name=task_name,
            rate_limit=row.get("rate_limit"),
            max_concurrent=row.get("max_concurrent"),
            max_retries=row.get("max_retries"),
            retry_backoff=row.get("retry_backoff"),
            timeout=row.get("timeout"),
            priority=row.get("priority"),
            paused=bool(row.get("paused", False)),
            updated_at=int(row.get("updated_at", 0)),
        )

    # ── Queues ─────────────────────────────────────────────────

    def list_queues(self) -> dict[str, QueueOverride]:
        out: dict[str, QueueOverride] = {}
        for key, raw in self._queue.list_settings().items():
            if not key.startswith(QUEUE_PREFIX):
                continue
            queue_name = key[len(QUEUE_PREFIX) :]
            out[queue_name] = self._row_to_queue(queue_name, _parse_json(raw))
        return out

    def get_queue(self, queue_name: str) -> QueueOverride | None:
        raw = self._queue.get_setting(QUEUE_PREFIX + queue_name)
        if not raw:
            return None
        return self._row_to_queue(queue_name, _parse_json(raw))

    def set_queue(self, queue_name: str, fields: dict[str, Any]) -> QueueOverride:
        _validate_queue_fields(fields)
        if not queue_name:
            raise ValueError("queue_name must not be empty")
        existing = self.get_queue(queue_name)
        merged: dict[str, Any] = {}
        if existing is not None:
            merged.update({k: v for k, v in asdict(existing).items() if v is not None})
            merged.pop("queue_name", None)
            merged.pop("updated_at", None)
        for k, v in fields.items():
            if v is None:
                merged.pop(k, None)
            else:
                merged[k] = v
        merged["updated_at"] = _now()
        self._queue.set_setting(
            QUEUE_PREFIX + queue_name, json.dumps(merged, separators=(",", ":"))
        )
        return self._row_to_queue(queue_name, merged)

    def clear_queue(self, queue_name: str) -> bool:
        return self._queue.delete_setting(QUEUE_PREFIX + queue_name)

    @staticmethod
    def _row_to_queue(queue_name: str, row: dict[str, Any]) -> QueueOverride:
        return QueueOverride(
            queue_name=queue_name,
            rate_limit=row.get("rate_limit"),
            max_concurrent=row.get("max_concurrent"),
            paused=bool(row.get("paused", False)),
            updated_at=int(row.get("updated_at", 0)),
        )

    # ── Apply (used at worker startup) ─────────────────────────

    def apply_task_overrides(self, configs: list[Any]) -> list[str]:
        """Mutate each :class:`PyTaskConfig` in ``configs`` with any matching
        task override. Returns a list of task names that are paused (so the
        caller can skip enqueuing them).
        """
        overrides = self.list_tasks()
        paused: list[str] = []
        for config in configs:
            override = overrides.get(config.name)
            if override is None:
                continue
            for field, value in override.as_patch().items():
                if hasattr(config, field):
                    setattr(config, field, value)
            if override.paused:
                paused.append(config.name)
        return paused

    def apply_queue_overrides(
        self, queue_configs: dict[str, dict[str, Any]]
    ) -> dict[str, dict[str, Any]]:
        """Merge queue overrides into ``queue_configs``. Returns the merged
        dict (a copy)."""
        merged: dict[str, dict[str, Any]] = {k: dict(v) for k, v in queue_configs.items()}
        for queue_name, override in self.list_queues().items():
            slot = merged.setdefault(queue_name, {})
            if override.rate_limit is not None:
                slot["rate_limit"] = override.rate_limit
            if override.max_concurrent is not None:
                slot["max_concurrent"] = override.max_concurrent
            if override.paused:
                slot["paused"] = True
        return merged

File: sdks/node/src/dashboard/overridesStore.ts
// Persistent task & queue runtime overrides. Operators tune task/queue
// behaviour from the dashboard; decorator/registration values stay the
// defaults and any override recorded here wins. Storage layout follows the
// cross-SDK contract:
//
// - `overrides:task:<task_name>` — JSON of overridden fields for that task
// - `overrides:queue:<queue_name>` — JSON of overridden fields for that queue
//
// Overrides are applied at worker startup; changes do not affect a running
// worker until it restarts (queue `paused` propagates live via pause/resume).

import type { TaskConfigInput } from "../native";
import { createLogger } from "../utils";

const TASK_PREFIX = "overrides:task:";
const QUEUE_PREFIX = "overrides:queue:";

const log = createLogger("dashboard");

/** The settings-KV surface both `Queue` and the native handle satisfy. */
export interface SettingsAccess {
  getSetting(key: string): string | null;
  setSetting(key: string, value: string): void;
  deleteSetting(key: string): boolean;
  listSettings(): Record<string, string>;
}

/** Fields an operator may override per task (cross-SDK field names). */
export const TASK_OVERRIDE_FIELDS = new Set([
  "rate_limit",
  "max_concurrent",
  "max_retries",
  "retry_backoff",
  "timeout",
  "priority",
  "paused",
]);

/** Fields an operator may override per queue. */
export const QUEUE_OVERRIDE_FIELDS = new Set(["rate_limit", "max_concurrent", "paused"]);

/** An operator-set task override (snake_case, matching the wire contract). */
export interface TaskOverride {
  task_name: string;
  rate_limit: string | null;
  max_concurrent: number | null;
  max_retries: number | null;
  retry_backoff: number | null;
  timeout: number | null;
  priority: number | null;
  paused: boolean;
  updated_at: number;
}

/** An operator-set queue override. */
export interface QueueOverride {
  queue_name: string;
  rate_limit: string | null;
  max_concurrent: number | null;
  paused: boolean;
  updated_at: number;
}

// ── Validation ──────────────────────────────────────────────────────────

function validateCommon(fields: Record<string, unknown>): void {
  const rateLimit = fields.rate_limit;
  if (rateLimit !== undefined && rateLimit !== null) {
    if (typeof rateLimit !== "string" || !rateLimit) {
      throw new Error("rate_limit must be a non-empty string like '100/m'");
    }
    // Cheap shape check; rate-limit parsing happens in the core.
    if (!rateLimit.includes("/")) {
      throw new Error("rate_limit must contain a unit, e.g. '10/s', '100/m', '3600/h'");
    }
  }
  validateInt(fields, "max_concurrent", 0);
  validateBool(fields, "paused");
}

function validateTaskFields(fields: Record<string, unknown>): void {
  const unknown = Object.keys(fields).filter((k) => !TASK_OVERRIDE_FIELDS.has(k));
  if (unknown.length > 0) {
    throw new Error(`unknown task override fields: ${unknown.sort().join(", ")}`);
  }
  validateCommon(fields);
  validateInt(fields, "max_retries", 0);
  validateNumber(fields, "retry_backoff", 0);
  validateInt(fields, "timeout", 1);
  validateInt(fields, "priority");
}

function validateQueueFields(fields: Record<string, unknown>): void {
  const unknown = Object.keys(fields).filter((k) => !QUEUE_OVERRIDE_FIELDS.has(k));
  if (unknown.length > 0) {
    throw new Error(`unknown queue override fields: ${unknown.sort().join(", ")}`);
  }
  validateCommon(fields);
}

function validateInt(fields: Record<string, unknown>, name: string, minimum?: number): void {
  const value = fields[name];
  if (value === undefined || value === null) {
    return;
  }
  if (typeof value !== "number" || !Number.isInteger(value)) {
    throw new Error(`${name} must be an integer`);
  }
  if (minimum !== undefined && value < minimum) {
    throw new Error(`${name} must be >= ${minimum}`);
  }
}

function validateNumber(fields: Record<string, unknown>, name: string, minimum?: number): void {
  const value = fields[name];
  if (value === undefined || value === null) {
    return;
  }
  if (typeof value !== "number" || !Number.isFinite(value)) {
    throw new Error(`${name} must be a number`);
  }
  if (minimum !== undefined && value < minimum) {
    throw new Error(`${name} must be >= ${minimum}`);
  }
}

function validateBool(fields: Record<string, unknown>, name: string): void {
  const value = fields[name];
  if (value !== undefined && value !== null && typeof value !== "boolean") {
    throw new Error(`${name} must be a boolean`);
  }
}

// ── Store ───────────────────────────────────────────────────────────────

function parseJsonObject(raw: string | null): Record<string, unknown> {
  if (!raw) {
    return {};
  }
  try {
    const data = JSON.parse(raw);
    return data && typeof data === "object" && !Array.isArray(data) ? data : {};
  } catch {
    log.warn(() => "overrides entry is not valid JSON; treating as empty");
    return {};
  }
}

/** CRUD for per-task and per-queue runtime overrides. */
export class OverridesStore {
  constructor(private readonly settings: SettingsAccess) {}

  // ── Tasks ─────────────────────────────────────────────────────────

  listTasks(): Map<string, TaskOverride> {
    const out = new Map<string, TaskOverride>();
    for (const [key, raw] of Object.entries(this.settings.listSettings())) {
      if (key.startsWith(TASK_PREFIX)) {
        const name = key.slice(TASK_PREFIX.length);
        out.set(name, rowToTask(name, parseJsonObject(raw)));
      }
    }
    return out;
  }

  getTask(taskName: string): TaskOverride | undefined {
    const raw = this.settings.getSetting(TASK_PREFIX + taskName);
    return raw ? rowToTask(taskName, parseJsonObject(raw)) : undefined;
  }

  /** Merge `fields` into the stored override; `null` values clear a field. */
  setTask(taskName: string, fields: Record<string, unknown>): TaskOverride {
    validateTaskFields(fields);
    if (!taskName) {
      throw new Error("task_name must not be empty");
    }
    const merged = this.mergeRow(this.settings.getSetting(TASK_PREFIX + taskName), fields);
    this.settings.setSetting(TASK_PREFIX + taskName, JSON.stringify(merged));
    return rowToTask(taskName, merged);
  }

  clearTask(taskName: string): boolean {
    return this.settings.deleteSetting(TASK_PREFIX + taskName);
  }

  // ── Queues ────────────────────────────────────────────────────────

  listQueues(): Map<string, QueueOverride> {
    const out = new Map<string, QueueOverride>();
    for (const [key, raw] of Object.entries(this.settings.listSettings())) {
      if (key.startsWith(QUEUE_PREFIX)) {
        const name = key.slice(QUEUE_PREFIX.length);
        out.set(name, rowToQueue(name, parseJsonObject(raw)));
      }
    }
    return out;
  }

  getQueue(queueName: string): QueueOverride | undefined {
    const raw = this.settings.getSetting(QUEUE_PREFIX + queueName);
    return raw ? rowToQueue(queueName, parseJsonObject(raw)) : undefined;
  }

  setQueue(queueName: string, fields: Record<string, unknown>): QueueOverride {
    validateQueueFields(fields);
    if (!queueName) {
      throw new Error("queue_name must not be empty");
    }
    const merged = this.mergeRow(this.settings.getSetting(QUEUE_PREFIX + queueName), fields);
    this.settings.setSetting(QUEUE_PREFIX + queueName, JSON.stringify(merged));
    return rowToQueue(queueName, merged);
  }

  clearQueue(queueName: string): boolean {
    return this.settings.deleteSetting(QUEUE_PREFIX + queueName);
  }

  private mergeRow(
    existingRaw: string | null,
    fields: Record<string, unknown>,
  ): Record<string, unknown> {
    const merged = parseJsonObject(existingRaw);
    delete merged.updated_at;
    for (const [key, value] of Object.entries(fields)) {
      if (value === null || value === undefined) {
        delete merged[key];
      } else {
        merged[key] = value;
      }
    }
    merged.updated_at = Date.now();
    return merged;
  }
}

function rowToTask(taskName: string, row: Record<string, unknown>): TaskOverride {
  return {
    task_name: taskName,
    rate_limit: asStringOrNull(row.rate_limit),
    max_concurrent: asNumberOrNull(row.max_concurrent),
    max_retries: asNumberOrNull(row.max_retries),
    retry_backoff: asNumberOrNull(row.retry_backoff),
    timeout: asNumberOrNull(row.timeout),
    priority: asNumberOrNull(row.priority),
    paused: row.paused === true,
    updated_at: asNumberOrNull(row.updated_at) ?? 0,
  };
}

function rowToQueue(queueName: string, row: Record<string, unknown>): QueueOverride {
  return {
    queue_name: queueName,
    rate_limit: asStringOrNull(row.rate_limit),
    max_concurrent: asNumberOrNull(row.max_concurrent),
    paused: row.paused === true,
    updated_at: asNumberOrNull(row.updated_at) ?? 0,
  };
}

const asStringOrNull = (v: unknown): string | null => (typeof v === "string" && v ? v : null);
const asNumberOrNull = (v: unknown): number | null =>
  typeof v === "number" && Number.isFinite(v) ? v : null;

// ── Worker-startup application ──────────────────────────────────────────

/**
 * Patch worker task configs with stored overrides. Tasks that only exist as
 * an override (registered without options) get a fresh config entry.
 * `retry_backoff` is interpreted as the retry base delay in seconds;
 * `timeout`/`priority` have no worker-config slot here and surface via the
 * dashboard's effective view only.
 */
export function applyTaskOverrides(
  configs: TaskConfigInput[],
  taskNames: Iterable<string>,
  store: OverridesStore,
): TaskConfigInput[] {
  const overrides = store.listTasks();
  if (overrides.size === 0) {
    return configs;
  }
  const byName = new Map(configs.map((config) => [config.name, { ...config }]));
  for (const name of taskNames) {
    const override = overrides.get(name);
    if (!override) {
      continue;
    }
    const config = byName.get(name) ?? { name };
    if (override.rate_limit !== null) {
      config.rateLimit = override.rate_limit;
    }
    if (override.max_concurrent !== null) {
      config.maxConcurrent = override.max_concurrent;
    }
    if (override.max_retries !== null) {
      config.maxRetries = override.max_retries;
    }
    if (override.retry_backoff !== null) {
      config.retryBaseDelayMs = Math.round(override.retry_backoff * 1000);
    }
    byName.set(name, config);
  }
  return [...byName.values()];
}

/** Merge queue overrides into worker queue configs (adding new queues). */
export function applyQueueOverrides(
  configs: Array<{ name: string; maxConcurrent?: number; rateLimit?: string }>,
  store: OverridesStore,
): Array<{ name: string; maxConcurrent?: number; rateLimit?: string }> {
  const overrides = store.listQueues();
  if (overrides.size === 0) {
    return configs;
  }
  const byName = new Map(configs.map((config) => [config.name, { ...config }]));
  for (const [name, override] of overrides) {
    const config = byName.get(name) ?? { name };
    if (override.rate_limit !== null) {
      config.rateLimit = override.rate_limit;
    }
    if (override.max_concurrent !== null) {
      config.maxConcurrent = override.max_concurrent;
    }
    byName.set(name, config);
  }
  return [...byName.values()];
}
